# 🏦 Project Overview

Welcome to your **Smart Finance Assistant** development journey! This notebook will evolve from basic CSV processing to a complete AI-powered finance application.

**Final Application Components:**
- 💬 **AI Chat Interface** - Financial advice personality
- 📊 **Data Analysis** - CSV transaction processing  
- 🔍 **RAG System** - Retrieval from financial documents
- 🛠️ **Custom Tools** - Calculators and utilities
- 🌐 **Gradio UI** - Professional web interface

**Development Approach**: Build progressively from foundation to advanced features, using AI collaboration throughout.

---

# 🚀 Getting Started: Foundation Setup

## Initial Setup
This cell installs the necessary libraries. In a Colab environment, you would uncomment the first line.

In [ ]:
# Uncomment the line below when running in Google Colab
# !pip install gradio pandas hands-on-ai

# Import core libraries
import pandas as pd
import numpy as np
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

print("📦 Core libraries loaded successfully!")
print(f"Pandas version: {pd.__version__}")

📦 Core libraries loaded successfully!
Pandas version: 2.2.2


## Hands-on-AI Configuration

Set up the hands-on-ai package for advanced features (chat, RAG, tools):

In [ ]:
import os

# Configure hands-on-ai server connection
os.environ['HANDS_ON_AI_SERVER'] = 'https://ollama.locollm.org'
os.environ['HANDS_ON_AI_MODEL'] = 'gemma3:4b'
os.environ['HANDS_ON_AI_API_KEY'] = 'Curtin2026ISYS20015002'

print("🔑 Hands-on-AI configured successfully!")

🔑 Hands-on-AI configured successfully!


The API key this semester is:  **isys2001-assignment-key**

## Connection Test

Test that everything is working correctly:

In [ ]:
from hands_on_ai.chat import get_response

# Test the connection to the hands-on-ai server
try:
    response = get_response("Hello! I'm building a Smart Finance Assistant.")
    print("✅ Hands-on-AI connection successful!")
    print(f"Response: {response}")
except Exception as e:
    print(f"❌ Connection issue: {e}")
    print("You can still work on the data processing foundation without this.")

✅ Hands-on-AI connection successful!
Response: That’s fantastic! Building a Smart Finance Assistant is a really exciting and useful project. I'm happy to help you in any way I can. 

To start, could you tell me a little more about what you're working on? For example:

*   **What is the overall goal of your assistant?** (e.g., budgeting, tracking expenses, investing, financial advice, etc.)
*   **What platform are you building it for?** (e.g., a mobile app, a web application, a voice assistant?)
*   **What technologies are you considering?** (e.g., Python, JavaScript, a specific database, integrations with financial APIs?)
*   **What are some of the key features you're envisioning?** (e.g., automatic transaction categorization, goal setting, spending reports, alerts?)

Even just a rough outline of what you’re thinking would be a great place to start.  Don’t worry about having everything figured out – this is a journey! 

Let's start brainstorming!  What's the first thing you wanted to t

# 🏗️ Foundation: Data Processing Skills

Before building advanced features, establish solid data processing foundations. This section focuses on CSV transaction analysis - the core of your finance assistant.

## Foundation Skill Checkpoint ✅

**Master these basics before advancing to chat/RAG/tools:**
- [ ] Load and clean CSV transaction data
- [ ] Handle real-world data issues (dollar signs, missing values)
- [ ] Calculate spending summaries by category  
- [ ] Generate business-appropriate insights
- [ ] Format output for professional presentation
- [ ] Test functions with various data scenarios

::: {.callout-tip}
## 🤖 AI Collaboration Strategy

For this foundation work, use AI to:
1. **Generate initial code** with specific business context
2. **Handle data cleaning** and validation
3. **Create professional formatting** for outputs
4. **Suggest business insights** from data patterns
5. **Help with testing** edge cases and error handling

**Remember**: You're directing AI like a junior developer - always review and improve their suggestions!
:::

## Sample Transaction Data Setup

Create or load sample transaction data to work with:

In [ ]:
# 🤖 AI Collaboration Opportunity:
# Ask AI to help you create realistic sample transaction data
# Include Australian businesses and various spending categories

# Sample data structure for testing
sample_transactions = {
    'Date': ['2024-08-01', '2024-08-02', '2024-08-03', '2024-08-04', '2024-08-05'],
    'Amount': ['$45.50', '$12.00', '$89.95', '$3.50', '-$25.00'],  # Note: includes $ signs and refund
    'Category': ['Groceries', 'Transport', 'Entertainment', 'Coffee', 'Refund'],
    'Description': ['Woolworths', 'Opal Card', 'Concert Tickets', 'Campus Cafe', 'Returned Item']
}

# Create DataFrame from sample data
df_sample = pd.DataFrame(sample_transactions)
print("📋 Sample transaction data created:")
print(df_sample)

📋 Sample transaction data created:
          Date   Amount       Category                     Description
0   2024-12-20  $850.00        Flights                  Flight to Bali
1   2024-12-20  $120.50         Dining  Airport Dinner - Qantas Lounge
2   2024-12-21  $250.00  Accommodation           Hotel Stay - Seminyak
3   2024-12-21   $45.75      Transport                   Taxi to Hotel
4   2024-12-22   $75.00     Activities                     Surf Lesson
5   2024-12-22   $30.00      Souvenirs              Local Craft Market
6   2024-12-23  $150.00         Dining                     Seafood BBQ
7   2024-12-23   $60.00       Shopping              Beachwear Shopping
8   2024-12-24   $95.20     Activities                 Snorkeling Tour
9   2024-12-24   $22.50         Coffee                Beach Cafe Latte
10  2024-12-25  $400.00  Accommodation         Return Flight to Sydney
11  2024-12-25   $88.00         Dining                 Farewell Dinner
12  2024-12-26  $180.00        Flights    

---

# 📊 Six-Step Development Methodology

Your notebook must demonstrate the six-step methodology with clear evidence of AI collaboration at each step.

## STEP 1: Understand the Problem

**🎯 Define Your Finance Problem**

In this section, clearly state your chosen finance problem in business terms.

::: {.callout-note}
## Problem Definition Template

*Example: "I want to help people understand their spending habits by analyzing their bank transaction exports. The system should identify spending patterns, highlight potential savings opportunities, and provide actionable financial insights in easy-to-understand language."*

**🤖 AI Prompt to Try:**
```
"Help me brainstorm specific personal finance problems that could be solved
with a CSV analysis tool. I'm interested in helping people with budgeting
and spending awareness. What are common financial challenges that data
analysis could address?"
```

**Your Problem Statement:**
I want to help travelers and international users better manage their travel expenses by automatically converting uploaded receipts, invoices, and expense reports into their preffered currency. The system should analyse uploaded expense data, apply accurate exchange rates, organise spending categories, and provide clear, easy to understand summaries of total costs and spending patterns. This will help users track their budgets more effectively, compare expenses across different countries, and make informed financial decisions while travelling.

## STEP 2: Identify Inputs and Outputs

**📥 Define Your Data Flow**

Clearly specify what data you'll work with and what insights you'll generate.

## Input/Output Analysis Template

**Inputs:**
- CSV file with transaction data (columns: Date, Amount, Category, Description)
- User preferences (spending goals, categories to focus on)
- Time period for analysis

**Outputs:**
- Total spending summary by category
- Average spending patterns
- Spending trend analysis
- Actionable savings recommendations
- Visual spending breakdown (future enhancement)

**🤖 AI Prompt to Try:**
```
"For a personal finance assistant that analyzes CSV transaction data,
what are the most valuable inputs I should collect and outputs I should
provide? Consider both technical data processing and business value
for the user."
```

**Your Input/Output Definition:**
The system will take the travel expenses, expense details, exchange rate data, and user preferences as inputs. It will then convert all the expenses into the user's preferred currency and generate outputs such as spending summaries, category breakdowns, budget tracking, exchange rate calculations, travel spending insights, and expense reports.

## STEP 3: Work the Problem by Hand

**✋ Manual Calculation Examples**

Total spending: 850 + 120.50 + 250 + 45.75 + 75 = $1341.25

Spending by category: flight= 850, Dining= 120.50, Accomodation= 250, Transport= 45.75, Activites= 75
Highest spending category= Flights at $850

Average transaction


## Example Business Calculation

Given this sample data:

| Date | Amount | Category | Description |
|------|--------|----------|-------------|
| 2024-08-01 | $45.50 | Groceries | Woolworths |
| 2024-08-02 | $12.00 | Transport | Opal Card |
| 2024-08-03 | $89.95 | Entertainment | Concert |
| 2024-08-04 | -$15.00 | Refund | Returned item |

**Manual Calculations:**
- Total Spending: $45.50 + $12.00 + $89.95 - $15.00 = $132.45
- By Category: Groceries $45.50, Transport $12.00, Entertainment $89.95
- Average Transaction: $132.45 ÷ 4 = $33.11
- Insight: Entertainment represents 67% of positive spending

**🤖 AI Prompt to Try:**
```
"Give me 3 realistic examples of Australian transaction data with different
scenarios (normal spending, refunds, large purchases). Show me how to manually
calculate meaningful financial insights from each scenario."
```

**Your Manual Examples:**


## STEP 4: Write Pseudocode

**📝 Plan Your Solution Logic**

Sketch the algorithm in plain English before coding.


## Pseudocode Template

```
FUNCTION analyze_spending_data(csv_file):
    // Step 1: Load and validate data
    - Read CSV file into DataFrame
    - Check for required columns (Date, Amount, Category, Description)
    - Handle missing or invalid data
    
    // Step 2: Clean financial data  
    - Remove dollar signs from Amount column
    - Convert to numeric values
    - Separate positive spending from refunds
    
    // Step 3: Calculate business metrics
    - Group transactions by category
    - Calculate totals, averages, counts per category
    - Identify spending patterns and outliers
    
    // Step 4: Generate insights
    - Calculate percentages of total spending
    - Identify top spending categories
    - Suggest potential savings opportunities
    
    // Step 5: Format for presentation
    - Create business-friendly output
    - Include currency formatting
    - Add explanatory text for insights
    
    RETURN formatted_analysis
```

**🤖 AI Prompt to Try:**
```
"Review my pseudocode for analyzing spending data from a CSV. Are there any
edge cases I'm missing? What business logic should I add to make this more
valuable for personal finance insights?"
```


**Your Pseudocode:**
[Write your detailed algorithm here]

## STEP 5: Convert to Python

**💻 Implementation with AI Collaboration**

Now implement your solution using AI assistance. Focus on creating professional, business-appropriate code.


## 🤖 Implementation Strategy

**Effective AI Prompts for Implementation:**
```
"I'm implementing a Smart Finance Assistant. Based on my pseudocode, please create
a Python function that [specific functionality]. The code should:
- Handle real-world CSV data issues (dollar signs, missing values)
- Include clear comments explaining business logic
- Use professional variable names
- Format output for business presentation
- Include basic error handling"
```

**Remember to critique and improve AI responses before using them!**
:::

### Foundation Data Processing Functions

---

# 📊 Six-Step Development Methodology

Your notebook must demonstrate the six-step methodology with clear evidence of AI collaboration at each step.

## STEP 1: Understand the Problem

**🎯 Define Your Finance Problem**

In this section, clearly state your chosen finance problem in business terms.

::: {.callout-note}
## Problem Definition Template

*Example: "I want to help people understand their spending habits by analyzing their bank transaction exports. The system should identify spending patterns, highlight potential savings opportunities, and provide actionable financial insights in easy-to-understand language."*

**🤖 AI Prompt to Try:**
```
"Help me brainstorm specific personal finance problems that could be solved
with a CSV analysis tool. I'm interested in helping people with budgeting
and spending awareness. What are common financial challenges that data
analysis could address?"
```

**Your Problem Statement:**
I want to help travelers and international users better manage their travel expenses by automatically converting uploaded receipts, invoices, and expense reports into their preffered currency. The system should analyse uploaded expense data, apply accurate exchange rates, organise spending categories, and provide clear, easy to understand summaries of total costs and spending patterns. This will help users track their budgets more effectively, compare expenses across different countries, and make informed financial decisions while travelling.

## STEP 2: Identify Inputs and Outputs

**📥 Define Your Data Flow**

Clearly specify what data you'll work with and what insights you'll generate.

## Input/Output Analysis Template

**Inputs:**
- CSV file with transaction data (columns: Date, Amount, Category, Description)
- User preferences (spending goals, categories to focus on)
- Time period for analysis

**Outputs:**
- Total spending summary by category
- Average spending patterns
- Spending trend analysis
- Actionable savings recommendations
- Visual spending breakdown (future enhancement)

**🤖 AI Prompt to Try:**
```
"For a personal finance assistant that analyzes CSV transaction data,
what are the most valuable inputs I should collect and outputs I should
provide? Consider both technical data processing and business value
for the user."
```

**Your Input/Output Definition:**
The system will take the travel expenses, expense details, exchange rate data, and user preferences as inputs. It will then convert all the expenses into the user's preferred currency and generate outputs such as spending summaries, category breakdowns, budget tracking, exchange rate calculations, travel spending insights, and expense reports.

## STEP 3: Work the Problem by Hand

**✋ Manual Calculation Examples**


## Example Business Calculation

Given this sample data:

| Date | Amount | Category | Description |
|------|--------|----------|-------------|
| 2024-08-01 | $45.50 | Groceries | Woolworths |
| 2024-08-02 | $12.00 | Transport | Opal Card |
| 2024-08-03 | $89.95 | Entertainment | Concert |
| 2024-08-04 | -$15.00 | Refund | Returned item |

**Manual Calculations:**
- Total Spending: $45.50 + $12.00 + $89.95 - $15.00 = $132.45
- By Category: Groceries $45.50, Transport $12.00, Entertainment $89.95
- Average Transaction: $132.45 ÷ 4 = $33.11
- Insight: Entertainment represents 67% of positive spending

**🤖 AI Prompt to Try:**
```
"Give me 3 realistic examples of Australian transaction data with different
scenarios (normal spending, refunds, large purchases). Show me how to manually
calculate meaningful financial insights from each scenario."
```

**Your Manual Examples:**
Total spending: 850+120.50+250+45.75+75=1341.25

Spending Categories:
Flights=850
dining=120.50
accomodation=250
transport=45.75
activities=75
Hihest spending category: flights= $850

Average Transaction:
Number of transactions=5
average transaction amount: 1341.25/5=268.25

## STEP 4: Write Pseudocode

**📝 Plan Your Solution Logic**

Sketch the algorithm in plain English before coding.


## Pseudocode Template

```
FUNCTION analyze_spending_data(csv_file):
    // Step 1: Load and validate data
    - Read CSV file into DataFrame
    - Check for required columns (Date, Amount, Category, Description)
    - Handle missing or invalid data
    
    // Step 2: Clean financial data  
    - Remove dollar signs from Amount column
    - Convert to numeric values
    - Separate positive spending from refunds
    
    // Step 3: Calculate business metrics
    - Group transactions by category
    - Calculate totals, averages, counts per category
    - Identify spending patterns and outliers
    
    // Step 4: Generate insights
    - Calculate percentages of total spending
    - Identify top spending categories
    - Suggest potential savings opportunities
    
    // Step 5: Format for presentation
    - Create business-friendly output
    - Include currency formatting
    - Add explanatory text for insights
    
    RETURN formatted_analysis
```

**🤖 AI Prompt to Try:**
```
"Review my pseudocode for analyzing spending data from a CSV. Are there any
edge cases I'm missing? What business logic should I add to make this more
valuable for personal finance insights?"
```


**Your Pseudocode:**
function analyze_spending_data (csv_file)

//step 1: load and validate data
-read travel expense file into dataframe
-check for required columns (date, amount, currency)
-handle missing or invalid data

//step 2: process expense data
-remove currency symbols from amounts
-convert amounts into numbers
-store cleaned travel expense data

//step 3: get currency conversion details
-ask user for preferred currency
-retrieve exchange rates
-validate exchange rate data

//step 4: convert travel expenses
-loop through each expense
-convert original amount into preferred currency
-save converted values

//step 5: analyze spending
-calculate total spending
-group expenses by category
-calculate average transaction amount
-identify highest spending category

//step 6: generate insights
-detect overspending areas
-create travel spending summary
-generate budget insights

//step 7: display results
-show converted expenses
-display spending summaries
-export final expense report

## STEP 5: Convert to Python

**💻 Implementation with AI Collaboration**

Now implement your solution using AI assistance. Focus on creating professional, business-appropriate code.


## 🤖 Implementation Strategy

**Effective AI Prompts for Implementation:**
```
"I'm implementing a Smart Finance Assistant. Based on my pseudocode, please create
a Python function that [specific functionality]. The code should:
- Handle real-world CSV data issues (dollar signs, missing values)
- Include clear comments explaining business logic
- Use professional variable names
- Format output for business presentation
- Include basic error handling"
```

**Remember to critique and improve AI responses before using them!**
:::

### Foundation Data Processing Functions

In [ ]:
def load_and_clean_transaction_data(file_path):
    required_columns = ['Date', 'Amount', 'Category', 'Description']

    try:
        if isinstance(file_path, str):
            df = pd.read_csv(file_path)
        else:
            df = pd.read_csv(file_path)

    except FileNotFoundError:
        raise FileNotFoundError(f"Business Error: The transaction file '{file_path}' was not found. Please ensure the correct path is provided.")
    except Exception as e:
        raise ValueError(f"Business Error: Could not load the transaction data. Please check the file format and content. Details: {e}")

    if not all(col in df.columns for col in required_columns):
        missing = [col for col in required_columns if col not in df.columns]
        raise ValueError(f"Business Error: The transaction data is missing essential columns: {', '.join(missing)}. Please ensure the CSV contains 'Date', 'Amount', 'Category', and 'Description'.")

    if 'Amount' in df.columns:
        try:
            df['Amount'] = df['Amount'].astype(str).str.replace(r'[$,€£]', '', regex=True).str.replace(',', '', regex=False)
            df['Amount'] = pd.to_numeric(df['Amount'], errors='coerce')
        except Exception as e:
            raise ValueError(f"Business Error: Failed to convert 'Amount' column to numbers. Please ensure amounts are in a valid numeric format (e.g., '$45.50' or '45.50'). Details: {e}")

    if 'Date' in df.columns:
        try:
            df['Date'] = pd.to_datetime(df['Date'], errors='coerce')
        except Exception as e:
            raise ValueError(f"Business Error: Failed to convert 'Date' column to datetime format. Please ensure dates are in a recognizable format (e.g., 'YYYY-MM-DD'). Details: {e}")

    initial_rows = len(df)
    df.dropna(subset=['Date', 'Amount', 'Category', 'Description'], inplace=True)
    if len(df) < initial_rows:
        print(f"Warning: {initial_rows - len(df)} rows with missing or invalid essential data were removed.")

    return df

In [ ]:
def analyze_spending_patterns(df):
    if df.empty:
        return {
            "message": "No transactions to analyze.",
            "total_spending": 0.0,
            "category_spending": {},
            "spending_percentages": {},
            "top_category": "N/A",
            "average_transaction_amount": 0.0,
            "insights": []
        }

    total_spending = df['Amount'].sum()

    category_spending = df.groupby('Category')['Amount'].sum().sort_values(ascending=False)

    positive_spending = df[df['Amount'] > 0]['Amount'].sum()
    spending_percentages = {}
    if positive_spending > 0:
        for category, amount in category_spending.items():
            if amount > 0:
                spending_percentages[category] = (amount / positive_spending) * 100
            else:
                spending_percentages[category] = 0.0

    spending_percentages = {k: v for k, v in sorted(spending_percentages.items(), key=lambda item: item[1], reverse=True)}

    top_category = category_spending.index[0] if not category_spending.empty else "N/A"

    positive_transactions = df[df['Amount'] > 0]['Amount']
    average_transaction_amount = positive_transactions.mean() if not positive_transactions.empty else 0.0

    insights = []
    if total_spending > 0:
        insights.append(f"Your total expenditure across all categories is ${total_spending:.2f}.")
        insights.append(f"The highest spending category is '{top_category}' accounting for ${category_spending.iloc[0]:.2f}.")
        if 'Flights' in category_spending and 'Accommodation' in category_spending:
            insights.append("Flights and Accommodation are typically the largest components of travel budgets. Consider booking in advance or looking for package deals to save.")
        elif 'Dining' in category_spending and 'Activities' in category_spending:
            insights.append("Significant spending on Dining and Activities suggests a focus on experiences. Look for local deals or cooking some meals to reduce costs.")
        if average_transaction_amount > 100:
            insights.append(f"Your average positive transaction amount is ${average_transaction_amount:.2f}. Monitoring larger individual expenses could help with budget management.")
    elif total_spending < 0:
        insights.append(f"Your net balance is positive due to refunds, with a total of ${total_spending:.2f}. Good job on managing returns!")
    else:
        insights.append("Your net spending is zero, indicating a balanced budget or transactions primarily consisting of refunds.")


    return {
        "message": "Spending analysis completed successfully.",
        "total_spending": total_spending,
        "category_spending": category_spending.to_dict(),
        "spending_percentages": spending_percentages,
        "top_category": top_category,
        "average_transaction_amount": average_transaction_amount,
        "insights": insights
    }

from io import StringIO
import pandas as pd

sample_transactions = {
    'Date': ['2024-12-20', '2024-12-20', '2024-12-21', '2024-12-21', '2024-12-22', '2024-12-22', '2024-12-23', '2024-12-23', '2024-12-24', '2024-12-24', '2024-12-25', '2024-12-25', '2024-12-26', '2024-12-26', '2024-12-27'],
    'Amount': ['$850.00', '$120.50', '$250.00', '$45.75', '$75.00', '$30.00', '$150.00', '$60.00', '$95.20', '$22.50', '$400.00', '$88.00', '$180.00', '$15.00', '$50.00'],
    'Category': ['Flights', 'Dining', 'Accommodation', 'Transport', 'Activities', 'Souvenirs', 'Dining', 'Shopping', 'Activities', 'Coffee', 'Accommodation', 'Dining', 'Flights', 'Transport', 'Groceries'],
    'Description': ['Flight to Bali', 'Airport Dinner - Qantas Lounge', 'Hotel Stay - Seminyak', 'Taxi to Hotel', 'Surf Lesson', 'Local Craft Market', 'Seafood BBQ', 'Beachwear Shopping', 'Snorkeling Tour', 'Beach Cafe Latte', 'Return Flight to Sydney', 'Farewell Dinner', 'Hotel Check-out', 'Souvenir Shopping', 'Airport Shuttle']
}
df_sample = pd.DataFrame(sample_transactions)

sample_csv_data_for_test = df_sample.to_csv(index=False)
test_file_path_for_analysis = StringIO(sample_csv_data_for_test)

try:
    cleaned_df_for_analysis = load_and_clean_transaction_data(test_file_path_for_analysis)
    analysis = analyze_spending_patterns(cleaned_df_for_analysis)
    print("\n✅ Spending analysis successful!")
    print("Total Spending: ", analysis['total_spending'])
    print("Category Spending: ", analysis['category_spending'])
    print("Spending Percentages: ", analysis['spending_percentages'])
    print("Top Category: ", analysis['top_category'])
    print("Average Transaction Amount: ", analysis['average_transaction_amount'])
    print("Insights:")
    for insight in analysis['insights']:
        print(f"- {insight}")
except Exception as e:
    print(f"\n❌ Error during spending analysis test: {e}")


✅ Spending analysis successful!
Total Spending:  2431.95
Category Spending:  {'Flights': 1030.0, 'Accommodation': 650.0, 'Dining': 358.5, 'Activities': 170.2, 'Transport': 60.75, 'Shopping': 60.0, 'Groceries': 50.0, 'Souvenirs': 30.0, 'Coffee': 22.5}
Spending Percentages:  {'Flights': np.float64(42.35284442525546), 'Accommodation': np.float64(26.727523180986452), 'Dining': np.float64(14.741257015974837), 'Activities': np.float64(6.998499146775222), 'Transport': np.float64(2.4979954357614265), 'Shopping': np.float64(2.467155985937211), 'Groceries': np.float64(2.0559633216143425), 'Souvenirs': np.float64(1.2335779929686055), 'Coffee': np.float64(0.9251834947264541)}
Top Category:  Flights
Average Transaction Amount:  162.13
Insights:
- Your total expenditure across all categories is $2431.95.
- The highest spending category is 'Flights' accounting for $1030.00.
- Flights and Accommodation are typically the largest components of travel budgets. Consider booking in advance or looking for 

In [ ]:
def generate_financial_recommendations(analysis_data):
    report_sections = []

    total_spending = analysis_data.get('total_spending', 0.0)
    message = analysis_data.get('message', 'No specific message.')
    report_sections.append(f"--- Financial Recommendations Report ---")
    report_sections.append(f"Status: {message}")
    report_sections.append(f"Overall Total Spending: ${total_spending:.2f}\n")

    category_spending = analysis_data.get('category_spending', {})
    spending_percentages = analysis_data.get('spending_percentages', {})
    top_category = analysis_data.get('top_category', 'N/A')
    average_transaction_amount = analysis_data.get('average_transaction_amount', 0.0)

    report_sections.append("1. Spending Pattern Observations:")
    if category_spending:
        report_sections.append(f"   - Your highest spending category is **{top_category}** at ${category_spending.get(top_category, 0.0):.2f}. This accounts for {spending_percentages.get(top_category, 0.0):.2f}% of your positive spending.")
        if top_category in ['Flights', 'Accommodation']:
            report_sections.append("     Recommendation: For large travel expenses like flights and accommodation, consider booking further in advance, looking for package deals, or exploring loyalty programs to reduce costs.")
        elif top_category == 'Dining':
            report_sections.append("     Recommendation: Dining is a common area for budget optimization. Try meal prepping, exploring budget-friendly local eateries, or taking advantage of happy hour deals.")
        elif top_category == 'Shopping':
            report_sections.append("     Recommendation: Review your shopping habits. Differentiate between needs and wants, and consider a '24-hour rule' for non-essential purchases to avoid impulse buying.")
        else:
            report_sections.append(f"     Observation: Further review of '{top_category}' spending could reveal specific areas for optimization.")

        if average_transaction_amount > 100:
            report_sections.append(f"   - Your average positive transaction is ${average_transaction_amount:.2f}. Monitoring these larger individual expenses can help manage overall budget.")
    else:
        report_sections.append("   - No detailed category spending to analyze.")

    report_sections.append("\n2. Potential Savings Opportunities:")
    if 'Flights' in category_spending or 'Accommodation' in category_spending:
        report_sections.append("   - **Travel Costs:** Given your spending on travel-related categories, significant savings could be achieved by comparing prices across different platforms, being flexible with dates, or using points/miles if applicable.")
    if 'Dining' in category_spending and spending_percentages.get('Dining', 0) > 10:
        report_sections.append("   - **Dining Out:** High dining expenses indicate an opportunity to save. Cooking at home more often, packing lunches, or limiting restaurant visits to special occasions can significantly reduce this category.")
    if 'Coffee' in category_spending and category_spending['Coffee'] > 20:
         report_sections.append("   - **Small Frequent Purchases:** Even small daily purchases like coffee can add up. Consider making coffee at home a few days a week.")

    initial_insights = analysis_data.get('insights', [])
    if initial_insights:
        report_sections.append("\n3. Additional Insights:")
        for insight in initial_insights:
            report_sections.append(f"   - {insight}")

    report_sections.append("\n--- Actionable Advice ---")
    report_sections.append("\n*   **Set a Budget:** Establish clear spending limits for your top categories like 'Flights' and 'Accommodation' to prevent overspending.")
    report_sections.append("*   **Track Expenses:** Continue tracking your expenses diligently to monitor progress and identify any new patterns.")
    report_sections.append("*   **Review Regularly:** Schedule a weekly or monthly review of your spending to stay on top of your financial goals.")
    report_sections.append("*   **Seek Alternatives:** For categories with high spending, actively look for more cost-effective alternatives.")

    return "\n".join(report_sections)


✅ Financial recommendations generated successfully!
--- Financial Recommendations Report ---
Status: Spending analysis completed successfully.
Overall Total Spending: $2431.95

1. Spending Pattern Observations:
   - Your highest spending category is **Flights** at $1030.00. This accounts for 42.35% of your positive spending.
     Recommendation: For large travel expenses like flights and accommodation, consider booking further in advance, looking for package deals, or exploring loyalty programs to reduce costs.
   - Your average positive transaction is $162.13. Monitoring these larger individual expenses can help manage overall budget.

2. Potential Savings Opportunities:
   - **Travel Costs:** Given your spending on travel-related categories, significant savings could be achieved by comparing prices across different platforms, being flexible with dates, or using points/miles if applicable.
   - **Dining Out:** High dining expenses indicate an opportunity to save. Cooking at home more

---

# 🌐 Advanced Features: Integrating AI Components

Once your foundation data processing is solid, integrate advanced AI features using hands-on-ai.

## Chat Interface Integration

In [ ]:
import os
!pip install hands-on-ai
from hands_on_ai.chat import get_response

def create_finance_chat_personality():
    system_prompt = (
        "You are 'Finny', a friendly, professional, and practical financial advisor AI within a Smart Finance Assistant. "
        "Your primary goal is to help users understand their spending habits, especially for travel expenses, "
        "offer actionable advice to manage finances, and identify savings opportunities. "
        "You can also assist with general currency conversion inquiries, but do not provide real-time exchange rates. "
        "Always be encouraging but realistic. Break down complex financial concepts into easy-to-understand language. "
        "Provide clear, concise, and helpful responses, leveraging the user's transaction analysis when available."
    )

    os.environ['HANDS_ON_AI_SYSTEM_PROMPT'] = system_prompt
    print("✅ Finny the Financial Advisor chatbot personality configured successfully via environment variable!")

create_finance_chat_personality()

try:
    test_query = "I'm planning a trip to Japan and need to manage my spending. How can I keep my travel costs down?"
    print(f"\nUser: {test_query}")
    response = get_response(test_query)
    print(f"Finny: {response}")
except Exception as e:
    print(f"❌ Error testing chatbot: {e}")
    print("Please ensure the hands-on-ai server connection is working.")

✅ Finny the Financial Advisor chatbot personality configured successfully via environment variable!

User: I'm planning a trip to Japan and need to manage my spending. How can I keep my travel costs down?
Finny: Okay, planning a trip to Japan is fantastic! It can be expensive, but with a little planning, you can definitely keep your costs down. Here’s a breakdown of strategies, broken down into categories, to help you manage your spending:

**1. Flights - The Biggest Expense**

* **Be Flexible with Dates:** This is *the* most impactful thing you can do.  Flying mid-week (Tuesday/Wednesday) and during the shoulder seasons (spring – late March/April – and autumn – October/November) are significantly cheaper than peak seasons like Golden Week (late April/early May) or New Year's.
* **Use Flight Comparison Websites:** Skyscanner, Google Flights, Kayak, Momondo – compare prices across different sites. Set up price alerts to be notified of drops.
* **Consider Budget Airlines:** While not alw

## RAG System for Financial Documents

In [ ]:
import pandas as pd

financial_documents = [
    {
        "id": "travel_budgeting",
        "content": (
            "Travellers should divide expenses into flights, "
            "accommodation, dining, transport, shopping, and activities."
        )
    },
    {
        "id": "flight_savings",
        "content": (
            "Booking flights early and travelling during off-peak "
            "seasons can reduce travel costs."
        )
    },
    {
        "id": "accommodation_savings",
        "content": (
            "Budget hotels, hostels, and shared accommodation "
            "can help reduce accommodation expenses."
        )
    },
    {
        "id": "currency_conversion",
        "content": (
            "Currency conversion helps travellers compare spending "
            "between different countries and currencies."
        )
    }
]

def retrieve_documents(user_question):
    results = []
    question = user_question.lower()
    for doc in financial_documents:
        content = doc["content"].lower()
        if (
            "flight" in question and "flight" in content
        ) or (
            "accommodation" in question and "accommodation" in content
        ) or (
            "currency" in question and "currency" in content
        ):
            results.append(doc["content"])
    return results

print("Testing Simple RAG System...\n")

question = (
    "How can I reduce spending on flights "
    "and accommodation?"
)

print(f"User Question: {question}\n")

documents = retrieve_documents(question)

print("Retrieved Documents:\n")

for doc in documents:
    print(f"- {doc}\n")

Testing Simple RAG System...

User Question: How can I reduce spending on flights and accommodation?

Retrieved Documents:

- Travellers should divide expenses into flights, accommodation, dining, transport, shopping, and activities.

- Booking flights early and travelling during off-peak seasons can reduce travel costs.

- Budget hotels, hostels, and shared accommodation can help reduce accommodation expenses.



## Custom Financial Tools

In [ ]:
class FinanceAgent:

    def __init__(self):
        self.tools = {}

    def register_tool(self, tool_name, tool_function):
        self.tools[tool_name] = tool_function
        print(f"✅ Tool '{tool_name}' registered successfully!")

    def use_tool(self, tool_name, *args):
        if tool_name in self.tools:
            return self.tools[tool_name](*args)
        else:
            return "❌ Tool not found."


finance_agent = FinanceAgent()

def savings_calculator(
    current_savings,
    monthly_contribution,
    target_amount
):
    if (
        current_savings < 0 or
        monthly_contribution <= 0 or
        target_amount <= 0
    ):
        return "❌ Invalid input values."

    if current_savings >= target_amount:
        return (
            f"✅ Savings goal already reached!"
        )

    remaining = (
        target_amount - current_savings
    )

    months_needed = (
        remaining / monthly_contribution
    )

    years = int(months_needed // 12)

    months = int(months_needed % 12)

    return (
        f"It will take approximately "
        f"{years} year(s) and "
        f"{months} month(s) "
        f"to reach your savings goal."
    )


finance_agent.register_tool(
    "savings_calculator",
    savings_calculator
)


result = finance_agent.use_tool(
    "savings_calculator",
    1000,
    200,
    5000
)

print("\nSavings Calculator Result:")
print(result)

✅ Tool 'savings_calculator' registered successfully!

Savings Calculator Result:
It will take approximately 1 year(s) and 8 month(s) to reach your savings goal.


## Gradio UI Integration

In [ ]:
import gradio as gr

def convert_currency(from_curr, to_curr, amount):
    try:
        if not from_curr or not to_curr:
            return "⚠️ Please enter both currencies (e.g. USD, EUR)."

        if amount <= 0:
            return "⚠️ Amount must be greater than 0."

        result = currency_converter_agent.convert(
            from_curr.upper(),
            to_curr.upper(),
            amount
        )

        return f"""
💱 Conversion Result
-------------------------
{amount} {from_curr.upper()} → {result} {to_curr.upper()}
"""
    except Exception as e:
        return f"❌ Error: {e}"

with gr.Blocks(title="Currency Converter Agent") as demo:
    gr.Markdown("# 💱 Currency Converter Agent")
    gr.Markdown("Convert currencies instantly using your AI-powered agent.")

    with gr.Row():
        from_currency = gr.Textbox(label="From Currency (e.g. USD)")
        to_currency = gr.Textbox(label="To Currency (e.g. EUR)")

    amount = gr.Number(label="Amount", value=1)

    convert_btn = gr.Button("Convert 💱")

    output = gr.Markdown()

    convert_btn.click(
        fn=convert_currency,
        inputs=[from_currency, to_currency, amount],
        outputs=output
    )

demo.launch()

It looks like you are running Gradio on a hosted Jupyter notebook, which requires `share=True`. Automatically setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://7ffd0b532446d644b0.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


---

# 🧪 STEP 6: Test with a Variety of Data

**🔍 Comprehensive Testing Strategy**

Create thorough tests for your Smart Finance Assistant to ensure it handles real-world scenarios.

::: {.callout-tip}
## 🤖 AI Collaboration for Testing

**Effective Testing Prompts:**
```
"Help me create comprehensive test cases for my finance assistant. Include:
- Normal transaction data
- Edge cases (refunds, large amounts, missing data)
- Invalid data scenarios (corrupted files, wrong formats)
- Business logic validation (spending calculations, recommendations)
Create assert statements to verify each scenario."
```
:::

## Foundation Function Tests

In [ ]:
import pandas as pd
from io import StringIO

def create_test_data():
    return StringIO("""
Date,Amount,Category,Description
2024-01-01,$50.00,Groceries,Woolworths
2024-01-02,$20.00,Coffee,Starbucks
2024-01-03,$100.00,Dining,Restaurant
""")

def test_load_data():
    print("🧪 Test 1: Loading Data")

    df = load_and_clean_transaction_data(create_test_data())

    assert not df.empty, "Data should not be empty"
    assert "Amount" in df.columns, "Missing Amount column"
    assert df["Amount"].dtype == "float64", "Amount should be numeric"

    print("✅ Test 1 Passed\n")

def test_analysis():
    print("🧪 Test 2: Spending Analysis")

    df = load_and_clean_transaction_data(create_test_data())
    analysis = analyze_spending_patterns(df)

    assert analysis["total_spending"] > 0, "Total spending should be > 0"
    assert "category_spending" in analysis, "Missing category breakdown"
    assert "top_category" in analysis, "Missing top category"

    print("✅ Test 2 Passed\n")

def test_recommendations():
    print("🧪 Test 3: Recommendations")

    df = load_and_clean_transaction_data(create_test_data())
    analysis = analyze_spending_patterns(df)
    report = generate_financial_recommendations(analysis)

    assert isinstance(report, str), "Report should be a string"
    assert len(report) > 0, "Report should not be empty"

    print("✅ Test 3 Passed\n")

def run_tests():
    print("\n🔍 SIMPLE TEST SUITE STARTED\n")

    test_load_data()
    test_analysis()
    test_recommendations()

    print("🎉 ALL TESTS PASSED SUCCESSFULLY")

run_tests()


🔍 SIMPLE TEST SUITE STARTED

🧪 Test 1: Loading Data
✅ Test 1 Passed

🧪 Test 2: Spending Analysis
✅ Test 2 Passed

🧪 Test 3: Recommendations
✅ Test 3 Passed

🎉 ALL TESTS PASSED SUCCESSFULLY


## Advanced Integration Tests

In [ ]:
import pandas as pd
from io import StringIO

# -----------------------------
# SAMPLE DATA
# -----------------------------
def sample_csv():
    return StringIO("""
Date,Amount,Category,Description
2024-01-01,$50.00,Groceries,Woolworths
2024-01-02,$20.00,Coffee,Starbucks
2024-01-03,$100.00,Dining,Restaurant
""")

# -----------------------------
# TEST 1: FULL WORKFLOW
# -----------------------------
def test_full_workflow():
    print("🧪 Test 1: Full Workflow")

    # 1. Load + clean
    df = load_and_clean_transaction_data(sample_csv())
    assert not df.empty, "CSV should load correctly"

    # 2. Analyze spending
    analysis = analyze_spending_patterns(df)
    assert analysis["total_spending"] > 0, "Analysis failed"

    # 3. Generate recommendations
    report = generate_financial_recommendations(analysis)
    assert isinstance(report, str) and len(report) > 0, "Report failed"

    print("✅ Workflow: CSV → Analysis → Recommendations OK\n")


# -----------------------------
# TEST 2: CHAT INTEGRATION
# -----------------------------
def test_chat():
    print("🧪 Test 2: Chat System")

    response = get_response("What am I spending too much on?")
    assert isinstance(response, str), "Chat response should be text"
    assert len(response) > 0, "Chat response is empty"

    print("✅ Chat system working\n")


# -----------------------------
# TEST 3: TOOL INTEGRATION
# -----------------------------
def test_tool():
    print("🧪 Test 3: Tool System")

    result = finance_agent.use_tool("savings_calculator", 1000, 100, 5000)
    assert result is not None, "Tool returned nothing"

    print("✅ Tool system working\n")


# -----------------------------
# TEST 4: ERROR HANDLING
# -----------------------------
def test_errors():
    print("🧪 Test 4: Error Handling")

    # Invalid CSV input
    try:
        load_and_clean_transaction_data(None)
    except Exception:
        print("✅ Handled invalid file input")

    # Invalid tool input
    try:
        finance_agent.use_tool("savings_calculator", "bad", None, None)
    except Exception:
        print("✅ Handled invalid tool input")

    print("✅ Error handling working\n")


# -----------------------------
# RUN ALL TESTS
# -----------------------------
def run_integration_tests():
    print("\n🔍 SIMPLE INTEGRATION TESTS STARTED\n")

    test_full_workflow()
    test_chat()
    test_tool()
    test_errors()

    print("🎉 ALL INTEGRATION TESTS PASSED")


run_integration_tests()


🔍 SIMPLE INTEGRATION TESTS STARTED

🧪 Test 1: Full Workflow
✅ Workflow: CSV → Analysis → Recommendations OK

🧪 Test 2: Chat System
✅ Chat system working

🧪 Test 3: Tool System
✅ Tool system working

🧪 Test 4: Error Handling
✅ Handled invalid file input
✅ Handled invalid tool input
✅ Error handling working

🎉 ALL INTEGRATION TESTS PASSED


---

# 📊 Project Completion Checklist

## Foundation Skills ✅
- [ ] **Data Processing**: CSV loading and cleaning functions work reliably
- [ ] **Analysis Functions**: Spending summaries calculate correctly
- [ ] **Business Insights**: Recommendations are relevant and actionable  
- [ ] **Error Handling**: Graceful handling of data issues
- [ ] **Testing**: Comprehensive test coverage for core functions
- [ ] **Documentation**: Clear AI collaboration documentation in diary

## Advanced Integration ✅
- [ ] **Chat Interface**: Finance advisor personality implemented
- [ ] **RAG System**: Document retrieval for financial guidance
- [ ] **Custom Tools**: At least one financial calculator/utility
- [ ] **Gradio UI**: Professional, user-friendly interface
- [ ] **Full Integration**: All components work together seamlessly

## Professional Standards ✅
- [ ] **Code Quality**: Professional, commented, maintainable code
- [ ] **Business Focus**: Clear connection to real finance problems
- [ ] **User Experience**: Interface suitable for non-technical users
- [ ] **AI Collaboration**: Extensive, well-documented AI usage
- [ ] **Testing**: Robust validation of all features

## Project Documentation ✅  
- [ ] **Developer's Diary**: Complete AI collaboration documentation
- [ ] **README**: Clear project description and usage instructions
- [ ] **GitHub**: Regular commits showing development progress
- [ ] **Reflection**: Thoughtful analysis of learning and challenges

---

# 🎯 Final Thoughts: Your Finance Assistant Journey

Congratulations on building your Smart Finance Assistant! This project represents a significant achievement in modern business programming:

**Technical Skills Developed:**
- AI-assisted development workflows
- Professional data processing with pandas
- Integration of multiple AI technologies
- User interface design with Gradio
- Comprehensive software testing

**Business Skills Developed:**  
- Financial data analysis and insights
- User-centered application design
- Professional documentation practices
- Iterative development methodology
- Critical evaluation of AI suggestions

**Professional Preparation:**
- Experience with industry-standard AI collaboration
- Portfolio-ready application development
- Understanding of business problem-solving with technology
- Documentation practices for workplace environments

**Your Smart Finance Assistant demonstrates your ability to direct AI tools toward meaningful business solutions - exactly the skill set that modern BIS graduates need for career success!**

---

*Remember to document all AI collaborations in your Developer's Diary and maintain regular GitHub commits throughout your development process.*
